lib/invoker.js

```diff
@@ -24,21 +25,53 @@ const invoke = (userFunction) => {
     };
 };
 
-module.exports = (userFunctionUri, port) => {
-    const userFunction = lookUpFunction(userFunctionUri);
-
-    const server = new grpc.Server();
-    server.addService(services.RiffService, {invoke: invoke(userFunction)});
+const executeInit = async (initHook, timeout, server) => {
+    logger('Calling $init hook (if any)');
+    return await guardWithTimeout(initHook, timeout)
+        .then(() => logger('$init hook successfully completed (or skipped if none)'))
+        .catch((err) => {
+            logger(`Error during $init hook execution: ${err.toString()}`);
+            logger('Shutting down now');
+            server.forceShutdown();
+            return Promise.reject('$init error');
+        });
+};
 
+const registerShutdown = (globals, destroyHook, timeout, server) => {
     const shutdown = () => {
         logger('Graceful shutdown started');
-        server.tryShutdown(() => logger('Graceful shutdown completed'));
+        logger('Calling $destroy hook (if any)');
+
+        guardWithTimeout(destroyHook, timeout)
+            .then(() => {
+                logger('$destroy hook successfully completed (or skipped if none)');
+                server.tryShutdown(() => logger('Graceful shutdown completed'));
+            })
+            .catch((err) => {
+                logger(`Error during $destroy hook execution: ${err.toString()}`);
+                logger('Shutting down now');
+                server.forceShutdown();
+            });
     };
-    process.on('SIGTERM', shutdown);
-    process.on('SIGINT', shutdown);
+    globals.process.once('SIGTERM', shutdown);
```

Comment:
process is a true global object. If the goal is to make the signals testable, we should instead create some other mechanism to detect and pass the signal which can be stubbed for tests.

Reply:
What's the drawback with the approach in this PR? What solution would you recommend instead?

Reply:
the draw back is two fold:
1. the names are confusing since `global` and `process` are globals
2. the process object isn't a mock of the global process, it's only an event emitter

I'd suggest consolidating the two signals into a single event to trigger a shutdown. That event could be represented as a promise and the promise can be consumed wherever needed.

Reply:
> I'd suggest consolidating the two signals into a single event to trigger a shutdown. That event could be represented as a promise and the promise can be consumed wherever needed.

I see how it would remove the need to introduce a custom process object.
But how would the event receiver exposed to tests? (Not that it matters much here as there seems to be no reliable way to validate that the shutdown hook is called after shutdown).